DEMO-1 quarto notebook (run workflow from pecan.xml) (#3531)
PecanProject/pecan: documentation/tutorials/Demo_1_Basic_Run/run_pecan.qmd

- If you have a `pecan.xml` file preconfigured, place it in your working directory and copy its file path. Assign this path to the **settings_path** variable.
- Alternatively, for a sample or initial run, you can use the example `pecan.xml` file available at: `pecan/documentation/tutorials/Demo_1_Basic_Run/pecan.xml`.

```{r load-settings}
settings_path <- here::here("/pecan/documentation/tutorials/Demo_1_Basic_Run/pecan.xml")
```

# 4. Prepare and Validate Settings

After specifying the path to your `pecan.xml` file, the next step involves reading and preparing these settings. PEcAn provides utilities to process and validate your configurations before any execution begins. 

- `PEcAn.settings::read.settings(settings_path)`: Parses the `pecan.xml` file, converting its contents into an R list object that PEcAn can work with.
- `PEcAn.settings::prepare.settings(settings)`: Further prepares and validates the settings, checking for missing required fields and ensuring consistency.

```{r read-prepare-settings}
# Read the settings from the pecan.xml file
settings <- PEcAn.settings::read.settings(settings_path)

# Prepare and validate the settings

> settings <- PEcAn.settings::prepare.settings(settings)

Reviewer: is this required if we are hard-coding the file paths?

Reviewer: obsolete after https://github.com/AritraDey-Dev/pecan/pull/6

```

# 5. Explore the Settings Object

Once the settings have been read and prepared, it is often useful to inspect the structure of the `settings` object. This object is an R list containing all parameters and configurations for the PEcAn workflow.

*   `str(settings)` displays the internal structure of the `settings` object. This shows how the settings are represented in R and is useful for debugging and verifying settings.

```{r explore-settings}
str(settings)
```

# 6. Create Output Directory

Before running the workflow, we need to ensure that the output directory specified in the settings exists. This directory will store all the model outputs, configuration files, and results from the simulation.

> **Note:** All outputs will be written to `settings$outdir`. You can change this directory in your `pecan.xml` file or by modifying `settings$outdir` in the notebook before running the workflow.

```{r create-outdir}

> dir.create(settings$outdir, recursive = TRUE, showWarnings = FALSE)

Reviewer: This throws error 

```
Warning message:
In dir.create(settings$outdir, recursive = TRUE) :
  cannot create dir '/data/workflows', reason 'No such file or directory'
```

Suggestions:
1. Don't suppress warnings here
2. get user to create a directory such as `demo_outdir`
3. add a line like settings$outdir <- here::here('demo_outdir')`
4. add `demo_outdir` to .gitignore

Reviewer: fixed in https://github.com/AritraDey-Dev/pecan/pull/6

```

> # 7. Prepare Meteorological Data

Reviewer: What is the goal here? Is it pedagogically useful to show how to copy the file to a new directory? Is it because running remote execution requires inputs to be in a single folder? 

The relative path could be hard coded into pecan.xml. 

If the goal is to look at the met data, you could show the path and perhaps use something like skimr::skim() to summarize the contents.

Author: For the Docker stack, it is needed; if we are not using the Docker stack, it is not needed. So, considering both scenarios, I believe it's fine to keep it.

Author: just to add there might be a small issue with the file path — if the files are just copied over to the dbfiles folder, it should work fine in both cases.

This section sets up the path to the meteorological data file that the model will use for the simulation. The meteorological data contains essential climate variables like temperature, precipitation, solar radiation, and humidity that drive the ecosystem model.

```{r prepare-met-directory}
dir.create("/data/dbfiles", recursive = TRUE, showWarnings = FALSE)
# Copy the meteorological data file to the dbfiles directory
system("cp /pecan/documentation/tutorials/Demo_1_Basic_Run/dbfiles/AMF_US-NR1_BASE_HH_23-5.2004-01-01.2004-12-31.clim /data/dbfiles/")
```

# 8. Write Model Configuration Files

This step generates the model-specific configuration files that will be used to run the ecosystem model. The process involves:

1. Disabling database operations: Set `settings$database <- NULL`
2. Disabling write operations: Set `settings$write <- FALSE`
3. Setting PFT posterior files: Point to the Plant Functional Type parameter files
4. Generating configs: Use `runModule.run.write.configs()`

```{r write-configs}

> # Disable database operations for this demo

Reviewer: If there is no database element in settings (`database == NULL`) then there is no need for write = FALSE. 

Is there a reason not to configure pecan.xml as desired (with no database element)?

Reviewer: this comment may have been misguided. I didn't realize settings$write was valid.

```

# 10. Extract Model Results and Prepare for Analysis

After the model simulation completes, we need to extract the results and prepare them for analysis. This involves:

1. Reading the run ID
2. Setting up output paths
3. Defining time period
4. Loading model output

```{r get-plot-vars}
runid <- as.character(read.table(paste(settings$outdir, "/run/", "runs.txt", sep = ""))[1, 1]) # Note: if you are using an xml from a run with multiple ensembles this line will provide only the first run id
outdir <- paste(settings$outdir, "/out/", runid, sep = "")
start.year <- as.numeric(lubridate::year(settings$run$start.date))
end.year <- as.numeric(lubridate::year(settings$run$end.date))
model_output <- PEcAn.utils::read.output(
  runid,
  outdir,
  start.year,
  end.year,
  variables = NULL,
  dataframe = TRUE,
  verbose = FALSE
)
available_vars <- names(model_output)[!names(model_output) %in% c("posix", "time_bounds")]
```

# 11. Display Available Model Variables

This section shows all the variables that are available in the model output. These variables represent different ecosystem processes and states that the model has simulated.

```{r show-variables, echo=FALSE}

> var_lookup <- read.csv("/pecan/documentation/tutorials/Demo_1_Basic_Run/extfiles/variable_lookup.csv", stringsAsFactors = FALSE)

Reviewer: Instead of adding a new file `variable_lookup.csv`, It would be better to pull this information from `PEcAn.utils::standard_vars` [base/utils/data/standard_vars.csv]. standard_vars contains all of these variables except year; 

the fields map thus: 
```
standard_vars   variable_lookup  
Variable        Variable.name
Long.name       Description
```

Reviewer: fixed in https://github.com/AritraDey-Dev/pecan/pull/6

---
title: "Running Ecosystem Simulations Using PEcAn"
author: "Aritra Dey"
format:
  pdf:
    toc: true
    number-sections: false
    fig-width: 10
    fig-height: 6
    fig-dpi: 300
---

# 1. Introduction {#introduction}

Welcome to this PEcAn workflow notebook! This notebook will guide you through running an ecosystem model using PEcAn's programmatic interface.

## 1.1 What Is PEcAn?

PEcAn (Predictive Ecosystem Analyzer) is a scientific workflow system designed to make ecosystem modeling more transparent, repeatable, and accessible. It helps researchers:

- Run ecosystem models with standardized inputs and outputs
- Perform uncertainty analysis on model parameters
- Compare model predictions with observations
- Share and reproduce scientific workflows

## 1.2 What This Notebook Does

This notebook demonstrates how to:

1. Set up and configure a PEcAn workflow
2. Run an ecosystem model simulation
3. Analyze and visualize the results

## 1.3 Prerequisites

Before running this notebook, make sure you have:

- All the PEcAn packages installed. You can install all PEcAn packages and their dependencies by running the following command in the root of your PEcAn repository:

```

> make install

Reviewer: I'd suggest installing from binaries. In a later revision (not this PR) we can consider making files available in the packages themselves so that we don't require cloning the repository and the demo can be truly "standalone"

```suggestion

# Enable repository from pecanproject
options(repos = c(
  pecanproject = 'https://pecanproject.r-universe.dev',
  CRAN = 'https://cloud.r-project.org'))
# Download and install PEcAn.all in R
install.packages('PEcAn.all')

```

Author: Also, I am not including this part in a separate code block, as it will run unnecessarily every time the notebook is rendered. This is meant to be a one-time download for the user.

---
title: "Running Ecosystem Simulations Using PEcAn"
author: "Aritra Dey"
format:
  pdf:
    toc: true
    number-sections: false
    fig-width: 10
    fig-height: 6
    fig-dpi: 300
---

# 1. Introduction {#introduction}

Welcome to this PEcAn workflow notebook! This notebook will guide you through running an ecosystem model using PEcAn's programmatic interface.

## 1.1 What Is PEcAn?

PEcAn (Predictive Ecosystem Analyzer) is a scientific workflow system designed to make ecosystem modeling more transparent, repeatable, and accessible. It helps researchers:

- Run ecosystem models with standardized inputs and outputs
- Perform uncertainty analysis on model parameters
- Compare model predictions with observations
- Share and reproduce scientific workflows

> ## 1.2 What This Notebook Does
> 
> This notebook demonstrates how to:
> 
> 1. Set up and configure a PEcAn workflow
> 2. Run an ecosystem model simulation
> 3. Analyze and visualize the results

Reviewer: ```suggestion
## 1.2 What This Notebook Does

This notebook demonstrates how to:

1. Set up and configure a PEcAn workflow
2. Run an ecosystem model simulation
3. Analyze and visualize the results

### The Scenario Being Modeled:

We are modeling carbon and productivity dynamics at the Niwot Ridge Forest AmeriFlux site ([US-NR1](https://ameriflux.lbl.gov/sites/siteinfo/US-NR1), a high-elevation temperate coniferous forest in Colorado. The model configuration uses the SIPNET process-based ecosystem model, parameterized with a temperate coniferous plant functional type (PFT).

The simulation is run for the full year 2004 (January 1 – December 31) using AmeriFlux LBL meteorological drivers from the Niwot Ridge site. The ensemble setup specifies one model run focusing on net primary productivity (NPP) as the target output variable.

This scenario is designed to be a minimal, reproducible example to demonstrate how to run SIPNET within the PEcAn workflow. In later steps, this same framework can be extended to include more ensemble members, additional PFTs, longer time periods, or alternative meteorological inputs.

This run is based on a study by [Moore et. al. (2007)](https://doi.org/10.1016/j.agrformet.2008.04.013) that uses SIPNET to understand the relationship between water and carbon balance at this site. 

```

```{r version-info}
PEcAn.all::pecan_version()
```

# 3. Install SIPNET v1.3.0

If you haven't already installed the SIPNET binary, you can do so by running the following code. This will download the SIPNET binary to `demo_outdir/sipnet` and make it executable.

> Note: The `demo_outdir` directory will be created in the root of your PEcAn installation (i.e., at `pecan/demo_outdir/`). This directory will contain the SIPNET binary as well as the output generated by PEcAn in this demo.

```{r download-sipnet}
# Download and install SIPNET v1.3.0
source(
  here::here(
    "documentation/tutorials/Demo_1_Basic_Run/download_sipnet.R"
  )
)
```

> Note: You can find the most recent version of the SIPNET binary at: [SIPNET GitHub Releases](https://github.com/PecanProject/sipnet/releases), but this notebook is designed to work with SIPNET v1.3.0.

# 4. Load PEcAn Packages

First, we need to load the PEcAn R packages. These packages provide all the functions we'll use to run the workflow.

```{r libraries}
# Load the PEcAn.all package, which includes all necessary PEcAn functionality
library("PEcAn.all")
```

> # 5. Load PEcAn Settings Files
> 
> This section handles the step of loading the PEcAn settings.
> 
> For a demo run, you can use the example `pecan.xml` file available at: `pecan/documentation/tutorials/Demo_1_Basic_Run/pecan.xml`.
> 
> 
> ```{r load-settings}
> settings_path <- here::here("documentation/tutorials/Demo_1_Basic_Run/pecan.xml")

Reviewer: ```suggestion
# 5. Load PEcAn Settings File

PEcAn uses an XML-based settings file (pecan.xml) to configure model runs. This file defines key information about the run including: PFT(s), site location, time period of the run, the location of input files and where outputs will be saved. Other settings outside the scope of this demo include the types of analyses that will be performed, how to connect to a database, and how to run it on a high performance computing cluster  (we are using the default single model run on a single computer). 

You can read more about the settings file in the ["PEcAn XML" chapter](https://pecanproject.github.io/pecan-documentation/develop/pecanXML.html#pecanXML) of the documentation.

There is an example `pecan.xml` that has been configured for this demonstration. You can find it at `pecan/documentation/tutorials/Demo_1_Basic_Run/pecan.xml`.

This is how PEcAn loads the settings file:

```{r load-settings}
settings_path <- here::here("documentation/tutorials/Demo_1_Basic_Run/pecan.xml")
```

# 8. Create Output Directory

Before running the workflow, we need to ensure that the output directory specified in the settings exists. This directory will store all the model outputs, configuration files, and results from the simulation.

> **Note:** All outputs will be written to `settings$outdir`. You can change this directory in your `pecan.xml` file or by modifying `settings$outdir` in the notebook before running the workflow. You can also put the `sipnet` executable anywhere you prefer, as long as you update `settings$model$binary`.

```{r create-outdir}
dir.create(settings$outdir, recursive = TRUE, showWarnings = FALSE)
```

# 9. Write Model Configuration Files

This step generates the model-specific configuration files that will be used to run the ecosystem model. The process involves:

1. Disabling database operations: Set `settings$database <- NULL`
2. Disabling write operations: Set `settings$write <- FALSE`
3. Setting PFT posterior files: Point to the Plant Functional Type parameter files

> 4. Generating configs: Use `runModule.run.write.configs()`

Reviewer: Updating to be consistent with text

```suggestion
1. Disabling database write operations because we are not using a database
2. Generating SIPNET configuration files using the `runModule.run.write.configs()` function.
```

```{r write-configs}
settings$database <- NULL # Disable database operations for this demo

> settings$write <- FALSE # Disable write operations

Reviewer: Remove this line given @AritraDey-Dev's comment "the line settings$database <- NULL does need to be present." (https://github.com/PecanProject/pecan/pull/3531/files#r2255308101):

```suggestion
settings$write <- FALSE # Disable database write operations
```

Author: for this `settings$database <- NULL`  is needed,but write we can remove.

```

- A valid `pecan.xml` configuration file or use the example provided: `pecan/documentation/tutorials/Demo_1_Basic_Run/pecan.xml`

## 1.4 How to Use This Notebook

1. Each section is clearly marked with a heading
2. Code chunks are provided with explanations
3. You can run the code chunks sequentially
4. Once you have successfully run the demo, you can modify parameters to configure new runs and analyses

**Objective:**

This demo illustrates how to run a basic PEcAn workflow using an R-based Quarto notebook. It will cover loading settings, writing model configuration files, and running model simulations. This approach provides a programmatic alternative to the web-based PEcAn interface for executing ecosystem models.

# 2. Install SIPNET v1.3.0

If you haven't already installed the SIPNET binary, you can do so by running the following code. This will download the SIPNET binary to `demo_outdir/sipnet` and make it executable.

> Note: The `demo_outdir` directory will be created in the root of your PEcAn installation (i.e., at `pecan/demo_outdir/`). This directory will contain the SIPNET binary as well as the output generated by PEcAn in this demo.

> ```{r download-sipnet, eval = FALSE}

Author: @dlebauer, is there any specific reason to set `eval = FALSE`? If someone renders the notebook directly, they would need to run this step separately. Also, in the case of a gh Action, we would need to add this as a separate step before running `quarto render run_pecan.qmd`. i believe it is better to remove `eval = FALSE`.

Reviewer: That makes sense. I wasn't considering CI or the use case of rendering without having run the code already.

# 7. Create Output Directory

Before running the workflow, we need to ensure that the output directory specified in the settings exists. This directory will store all the model outputs, configuration files, and results from the simulation.

> **Note:** All outputs will be written to `settings$outdir`. You can change this directory in your `pecan.xml` file or by modifying `settings$outdir` in the notebook before running the workflow. You can also put the `sipnet` executable anywhere you prefer, as long as you update `settings$model$binary`.

```{r create-outdir}
dir.create(settings$outdir, recursive = TRUE, showWarnings = FALSE)
```

# 8. Prepare Meteorological Data

This section sets up the path to the meteorological data file that the model will use for the simulation. The meteorological data contains essential climate variables like temperature, precipitation, solar radiation, and humidity that drive the ecosystem model.

```{r prepare-met-directory}
dir.create(here::here("dbfiles"))
# Copy the meteorological data file to the dbfiles directory
file.copy(
  here::here("documentation/tutorials/Demo_1_Basic_Run/dbfiles/AMF_US-NR1_BASE_HH_23-5.2004-01-01.2004-12-31.clim"),

>   here::here("dbfiles/")

Reviewer 2: I don't follow why this is being copied instead of just used from where it sits

(Also you might consider using models/sipnet/inst/niwot.clim rather than adding another clim file to the repo)

Author: The same concern was also mentioned by David above. I previously encountered a path issue while running this, likely due to a mistake on my part, but I’m no longer facing that problem — so I believe it can be removed.

For the future, I think it would be helpful to include a dedicated code block to assist users in downloading AmeriFlux data sources.

Also, @infotroph, I feel like` models/sipnet/inst/niwot.clim `is a different dataset from the one I’m currently using, i.e.,` Demo_1_Basic_Run/dbfiles/AMF_US-NR1_BASE_HH_23-5.2004-01-01.2004-12-31.clim.` Could you please confirm this?

Reviewer 2: >  I feel like models/sipnet/inst/niwot.clim is a different dataset 

It's from the same site, but includes a longer time period (late 1998 through end of 2005) at a coarser timestep (12 hour) whereas the AMF file contains only 2004 at a half-hour timestep. For demo purposes (and for many real modeling questions too!) the lower time resolution won't be a problem.

Author: then would it be better to mention that the user can also use niwot.clim with a proper description, or should we directly replace the ameriflux dataset with niwot.clim?

Author: Graph plotted using ameriflux dataset:
<img width="890" height="549" alt="Screenshot from 2025-08-09 20-13-31" src="https://github.com/user-attachments/assets/7309cb13-a9c6-48d9-a793-17ceb5ad146c" />
<img width="889" height="460" alt="Screenshot from 2025-08-09 20-13-25" src="https://github.com/user-attachments/assets/4e4b8b29-f969-4972-ae1f-61c890019b72" />

Graphs plotted using niwot.clim dataset:
<img width="881" height="488" alt="Screenshot from 2025-08-09 20-12-26" src="https://github.com/user-attachments/assets/69ef82e4-7cf1-4e6c-b3de-55dc28f63797" />

<img width="799" height="455" alt="Screenshot from 2025-08-09 20-12-16" src="https://github.com/user-attachments/assets/5c72eb19-a275-4ccc-ac76-ef0b0a61ac1e" />

@infotroph From these graphs, it is clear that there will be differences in the outputs. Therefore, I think we should stick to the AmeriFlux dataset, which is also used by the web version. Later (in a separate PR), I’m considering adding a separate block in the notebook that will allow users to directly download the AmeriFlux dataset.

)
```

# 9. Write Model Configuration Files

This step generates the model-specific configuration files that will be used to run the ecosystem model. The process involves:

1. Disabling database operations: Set `settings$database <- NULL`
2. Disabling write operations: Set `settings$write <- FALSE`
3. Setting PFT posterior files: Point to the Plant Functional Type parameter files
4. Generating configs: Use `runModule.run.write.configs()`

```{r write-configs}
settings$database <- NULL # Disable database operations for this demo

> settings$write <- FALSE # Disable write operations

Reviewer 2: Why not just set them this way in pecan.xml? Learning about database settings doesn't seem essential for demo 1

Author: Based on what I observed while rendering the notebook, it works fine even if I don't set `settings$write <- FALSE.`
However, the line `settings$database <- NULL `does need to be present. Ideally, if the database config is not provided in the settings file, it should default to NULL. But that doesn't seem to be happening(possibly due to a bug in either read.settings or prepare.settings).So, for the sake of consistency and to avoid errors, I’ve explicitly added this line.

Reviewer 2: Looks like `prepare.settings` adds a `settings$database$dbfiles` -- does it actually hurt anything to let that stay?

Author: Yes, prepare.settings adds `settings$database$dbfiles` after it is checked, but since we are specifying the path to met path separately in our pecan.xml before doing prepare.settings, I don’t think it causes any issues.

Author: Also, just to add, we have nothing to do with the database configuration for this workflow, as it is running without a database connection.

Reviewer: I've suggested some changes for clarity (removing unnecessary steps, and outdated comments) in https://github.com/PecanProject/pecan/pull/3531/files#r2289521197 and https://github.com/PecanProject/pecan/pull/3531/files#r2289518631

  ylab = "LAI (m2 m-2)",
  main = "Leaf Area Index Over Time"
)
lines(model_output$posix, model_output$AbvGrndWood, col = "brown")
legend("topright", legend = c("LAI", "Above Ground Wood"), col = c("darkgreen", "brown"), lty = 1)
```

# 14. Conclusion

This notebook demonstrated how to set up, run, and analyze a PEcAn ecosystem model workflow programmatically. You can now modify parameters, try different models, or extend the analysis as needed.

Try editing the `pecan.xml` file. Give it a new name and update the **settings_path** variable at the beginning of this Demo to point to the new file. See how the changes affect the model output!

# 15. Clean Up Workflow Output (Optional)

If you want to remove all files and directories created by this workflow and start fresh, you can run the following code. This will delete the entire output directory specified in your settings. **Use with caution!**

```{r cleanup}
# WARNING: This will permanently delete all workflow output files!
# Uncomment the line below to enable cleanup.
# fs::dir_delete(settings$outdir)
```

# 16. Session Info

This section prints your R session information for reproducibility.

```{r session-info}
sessionInfo()

> ```

Reviewer 2: I recommend moving this to the top, so that it's available for debugging even if the notebook stops with an error before it gets to the end, and also calling `PEcAn.all::pecan_version()` in the same block.

Reviewer: I agree that this is useful for debugging, but I’d rather keep it at the end for tutorial readability. For beginners, leading with session metadata detracts from the pedagogical objectives before they’ve run or understood the material.

Author: When we render the notebook, it’s going to appear in the final document anyway, so placing it at the end or at the top shouldn’t cause any issues,but yes keeping it at the end makes more sense for a novice user. Also, for debugging, I feel the rendered block outputs are enough to understand what’s happening, and adding `PEcAn.all::pecan_version()` at the top definitely makes sense. WDYT, @infotroph?
